### PR TITLE
asserts: parse plug-names/slot-names constraints 

### PR DESCRIPTION
**Pull request by pedronis**
This introduces format 4 for snap-declaration.

### CODE REVIEW

File: asserts/ifacedecls.go [hunk 1]
```diff
@@ -501,6 +504,88 @@ func compileDeviceScopeConstraint(cMap map[string]interface{}, context string) (
...
+	if !ok {
+		return nil, fmt.Errorf("%s constraint entry must be a regexp or special $ value", whichName)
+	}
+	if strings.HasPrefix(s, "$") {
+		if !validSpecialNameConstraint.MatchString(s) {
+			return nil, fmt.Errorf("%s constraint entry special value %q is invalid", whichName, s)
+		}
+		return specialNameMatcher{special: s}, nil
+	}
+	rx, err := regexp.Compile("^(" + s + ")$")
```
Comment (bboozzoo) [inline asserts/ifacedecls.go]: Do we allow a full Go regex or should it be more conservative and use `CompilePOSIX()`?

Should we complain about `\s`, `' '`, `\\`?

Comment (pedronis) [reply asserts/ifacedecls.go]: we have allowed full Go regexes so far in the same context see other code here.  We haven't done a lot of extra checks so far. Our worried about people matching for spaces given that this must be names?

Comment (pedronis) [reply asserts/ifacedecls.go]: For context this is similar preexisting code:

https://github.com/snapcore/snapd/blob/master/asserts/ifacedecls.go#L240

Comment (bboozzoo) [reply asserts/ifacedecls.go]: Yes, like accidentally creating a regexp that can't really match a valid slot or plug name. Although, I see now there's similar code for attribute constraints, so probably not worth the extra complexity.

File: asserts/ifacedecls.go [hunk 2]
```diff
@@ -501,6 +504,88 @@ func compileDeviceScopeConstraint(cMap map[string]interface{}, context string) (
...
+}
+
+// Check checks whether name doesn't match the constraints.
+func (nc *NameConstraints) Check(whichName, name string, special map[string]string) error {
+	for _, m := range nc.matchers {
+		if err := m.match(name, special); err == nil {
+			return nil
+		}
+	}
+	return fmt.Errorf("%s %q does not match constraints", whichName, name)
```
[the remark this answers is not in the document]
Comment (pedronis) [reply asserts/ifacedecls.go]: We get that from higher levels, but then by its nature interfaces/policy eats most of these errors anyway.

File: asserts/ifacedecls.go (hunk 1 above)
Comment (pedronis) [reply asserts/ifacedecls.go]: I could look for explicit whitespace, more than that seems overkill complicated, there is a lot of ways to write a regexp (for better or worse).

Comment (bboozzoo) [reply asserts/ifacedecls.go]: Yeah, probably not worth to dwell on it.

File: asserts/ifacedecls.go [hunk 3]
```diff
@@ -501,6 +504,88 @@ func compileDeviceScopeConstraint(cMap map[string]interface{}, context string) (
 	}, nil
 }
 
+type nameMatcher interface {
+	match(name string, special map[string]string) error
+}
+
+var (
+	validSpecialNameConstraint = regexp.MustCompile("^\\$[A-Z][A-Z0-9_]*$")
```
[the remark this answers is not in the document]
Comment (jdstrand) [reply asserts/ifacedecls.go]: Yes, that would be helpful.

File: asserts/ifacedecls.go [hunk 4]
```diff
@@ -501,6 +504,88 @@ func compileDeviceScopeConstraint(cMap map[string]interface{}, context string) (
...
+		return nil, fmt.Errorf("%s constraints must be a list of regexps and special $ values", whichName)
+	}
+	matchers := make([]nameMatcher, 0, len(l))
+	for _, nm := range l {
+		matcher, err := compileNameMatcher(whichName, nm)
+		if err != nil {
+			return nil, err
+		}
+		matchers = append(matchers, matcher)
+
```
Comment (jdstrand) [inline asserts/ifacedecls.go]: Extra newline?

File: asserts/ifacedecls.go [hunk 5]
```diff
@@ -501,6 +504,88 @@ func compileDeviceScopeConstraint(cMap map[string]interface{}, context string) (
...
+	return nil
+}
+
+// NameConstraints implements a set of constraints on the names of slots or plugs.
+type NameConstraints struct {
+	matchers []nameMatcher
+}
+
+func compileNameConstraints(whichName string, constraints interface{}) (*NameConstraints, error) {
+
```
Comment (jdstrand) [inline asserts/ifacedecls.go]: Extra newline?

File: asserts/ifacedecls.go (hunk 1 above)
Comment (pedronis) [reply asserts/ifacedecls.go]: @bboozzoo I did this https://github.com/snapcore/snapd/pull/7812/commits/1ceddf926d8062ab9092b706d09bf8cab2f13d0f

File: asserts/ifacedecls.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply asserts/ifacedecls.go]: done

**Pull request merged**